### ISSUE TITLE
Panic recovery middleware should be able to respond with status code

### ISSUE THREAD

**Issue by ifraixedes** [go-swagger/go-swagger#1056]
## Problem statement

I've noticed that because of the order of [the `WriteHeader` call in the response template](https://github.com/go-swagger/go-swagger/blob/master/generator/templates/server/responses.gotmpl#L149) if there is an error in the `producer.Produce`, although unlikely, the panic recover middleware doesn't have the possibility to respond with a status code (e.g. 500) because the header has been already written.

To state this is easier to see a generated code, for example the _[get by id response of the examples which are in this repo](https://github.com/go-swagger/go-swagger/blob/master/examples/generated/restapi/operations/pet/get_pet_by_id_responses.go#L48)_

Do I miss anything or you think, like me, that's better to move the call to `WriteHeader` to the end of the function?

## Environment
swagger version: 0.10.0
go version: 1.8
OS:  Linux

**Comment by casualjim** [comment]
When you write to a response writer for the body go will flush the headers and the body. So if we want to return a status code then we have to do that before we start writing to the response writer, which happens in the producer.Produce call
I think this could be fixed in the panic middleware by having a buffered response that gets flushed at the end or when its buffer overflows perhaps?

**Comment by ifraixedes** [comment]
> When you write to a response writer for the body go will flush the headers and the body. So if we want to return a status code then we have to do that before we start writing to the response writer, which happens in the producer.Produce call

I missed that point, my simple test to confirm my initial thoughts have been to call to `panic` just after the [`rw.WriteHeader(200)`](https://github.com/go-swagger/go-swagger/blob/master/examples/generated/restapi/operations/pet/get_pet_by_id_responses.go#L48) and have basic panic recover middleware which sends status 500.

Then I confirmed my thoughts of the response status code that the client gets is 200 rather than 500 and a message in the console saying something like that the header has been written more than once.

When I basically changed the `panic` call before of the `w.WriteHeader(200)`, the client got a response with status 500. 

However, I think that you're right, because it may panic after starting to write the response body but before finishing.

> I think this could be fixed in the panic middleware by having a buffered response that gets flushed at the end or when its buffer overflows perhaps?

I don't understand what you exactly suggested it. Could you elaborate how a panic middleware could avoid such thing without making any change in the default template used by the generator?

Many thanks!

**Comment by casualjim** [comment]
you can implement another response writer that wraps the default responsewriter
where you wrap the Write method that writes to a buffer with a known size, and you postpone flushing until the buffer. That would then mean you have an opportunity to changing the status code.
Of course there is a performance impact with this approach but it does allow for capturing some more status codes
